### PR TITLE
fix: delete OS disk when deleting VM

### LINKED ISSUE #201: OS Disk not deleted with Machine
/kind bug

**What steps did you take and what happened:**
Create a machine with role node. Delete it. Wait for all deletion operations to complete. Recreate it. There will be a disk conflict, because the previous OS disk was detached but not deleted with the VM.

**What did you expect to happen:**
OS disk should be removed with VM. 

**Anything else you would like to add:**
Fix is in https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/201

**Environment:**

- cluster-api-provider-azure version: commit 9d988a5a7025840b0842904a776f77dc91ae0066
- Kubernetes version: (use `kubectl version`): 1.14.1
- OS (e.g. from `/etc/os-release`): Ubuntu 18.04

### PR DESCRIPTION
**Pull request by alexeldeib**
**What this PR does / why we need it**:
Remove OS disks upon VM deletion, since Azure doesn't do this automatically.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #201

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/cloud/azure/services/disks/disks.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+	Name string
+}
+
+// Get on disk is currently no-op. OS disks should only be deleted and will create with the VM automatically. 
+func (s *Service) Get(ctx context.Context, spec v1alpha1.ResourceSpec) (interface{}, error) {
+	return Spec{}, nil
+}
+
+// Reconcile on disk is currently no-op. OS disks should only be deleted and will create with the VM automatically. 
+func (s *Service) Reconcile(ctx context.Context, spec v1alpha1.ResourceSpec) error {
```
Comment (alexeldeib) [inline pkg/cloud/azure/services/disks/disks.go]: Didn't want to introduce unnecessary ops, don't love this.

File: pkg/cloud/azure/actuators/machine/reconciler.go [hunk 2]
```diff
@@ -210,6 +213,14 @@ func (r *Reconciler) Delete(ctx context.Context) error {
 		return errors.Wrapf(err, "Unable to delete network interface")
 	}
 
+	OSDiskSpec := &disks.Spec{
+		Name: fmt.Sprintf("%s_OSDisk", r.scope.Machine.Name),
```
Comment (awesomenix) [inline pkg/cloud/azure/actuators/machine/reconciler.go]: Can you please move this to pkg/azure/defaults.go, and also move the other hardcoded place virtualmachines.go as well

File: pkg/cloud/azure/actuators/machine/reconciler.go [hunk 3]
```diff
@@ -66,6 +68,7 @@ func NewReconciler(scope *actuators.MachineScope) *Reconciler {
 		networkInterfacesSvc:  networkinterfaces.NewService(scope.Scope),
 		virtualMachinesSvc:    virtualmachines.NewService(scope.Scope),
 		virtualMachinesExtSvc: virtualmachineextensions.NewService(scope.Scope),
+		disksSvc: disks.NewService(scope.Scope),
```
Comment (awesomenix) [inline pkg/cloud/azure/actuators/machine/reconciler.go]: may not be formatted.

File: pkg/cloud/azure/services/disks/mock_disks/BUILD.bazel [hunk 4]
```diff
@@ -0,0 +1,25 @@
+load("@io_bazel_rules_go//go:def.bzl", "go_library")
+load("@bazel_gomock//:gomock.bzl", "gomock")
+
+gomock(
+    name = "mocks",
+    out = "disks_mock.go",
```
Comment (awesomenix) [inline pkg/cloud/azure/services/disks/mock_disks/BUILD.bazel]: can you please checkin this file as well

Comment (alexeldeib) [reply pkg/cloud/azure/services/disks/mock_disks/BUILD.bazel]: added

File: pkg/cloud/azure/actuators/machine/reconciler.go (hunk 3 above)
Comment (alexeldeib) [reply pkg/cloud/azure/actuators/machine/reconciler.go]: fixed, dunno why `fmt` didn't hit this automatically on file save.

**Comment by justaugustus** [comment]
Thanks for this, Ace!
/lgtm
/approve

**Pull request merged**